### PR TITLE
rebuild [WIP]

### PR DESCRIPTION
**Pull request by SimonDanisch**
Refactor GLWindow, modernize and move more functionality from GLVisualize here.
Introduce modular window rendering.

### CODE REVIEW

File: src/callbacks.jl [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    GLFW.SetWindowCloseCallback(window, (window,) -> push!(s, false))
+    s
+end
+"""
+Size of the window. Must not correlate to the real pixel size.
+This is why there is also framebuffer_size.
+returns `Signal{Vec{2,Int}}`
+[GLFW Docs](http://www.glfw.org/docs/latest/group__window.html#gaaca1c2715759d03da9834eac19323d4a)
+"""
+function window_size(window, s::Signal{Vec{2,Int}}=Signal(Vec{2,Int}( GLFW.GetWindowSize(window))))
```
Comment (sjkelly) [inline src/callbacks.jl]: With a signature on `window` this can use `Base.size`

File: src/callbacks.jl [hunk 2]
```diff
@@ -0,0 +1,150 @@
+"""
+Returns a signal, which is true as long as the window is open.
+returns `Signal{Bool}`
+[GLFW Docs](http://www.glfw.org/docs/latest/group__window.html#gaade9264e79fae52bdb78e2df11ee8d6a)
+"""
+function window_close(window, s::Signal{Bool}=Signal(true))
```
Comment (sjkelly) [inline src/callbacks.jl]: Same here, but for `Base.close`

File: src/callbacks.jl [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    end)
+    s
+end
+"""
+Registers a callback for keyboard unicode input.
+returns an `Signal{Vector{Char}}`,
+containing the pressed char. Is empty, if no key is pressed.
+[GLFW Docs](http://www.glfw.org/docs/latest/group__input.html#ga1e008c7a8751cea648c8f42cc91104cf)
+"""
+function unicode_input(window, s::Signal{Vector{Char}}=Signal(Char[]))
```
Comment (sjkelly) [inline src/callbacks.jl]: I guess a lot of this is to be 0.5 idiomatic since there are anonymous functions there?

File: src/screen.jl [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    ]
+end
+function SimpleRectangle{T}(position::Vec{2,T}, width::Vec{2,T})
+    SimpleRectangle{T}(position..., width...)
+end
+function createwindow(
+        name::AbstractString, w, h;
+        debugging = false,
+        major = 3,
+        minor = 2,# this is what GLVisualize needs to offer all features
```
Comment (sjkelly) [inline src/screen.jl]: yusss

File: src/callbacks.jl (hunk 2 above)
Comment (SimonDanisch) [inline src/callbacks.jl]: well these are callbacks, which shouldn't really be used themselves.... It diverts from Base's semantic of `close(obj)-> obj is closed`, vs `close(window, signal) -> signal which indicates if window is open`

File: src/callbacks.jl (hunk 3 above)
Comment (SimonDanisch) [inline src/callbacks.jl]: I tried to use functors as well, but I didn't find a way to make them work with cfunction.
In the end I did some benchmarks and found out that the overhead from clojures is not measurable compared to the event overhead, so I decided it's worth to make it the way we'd like to have it in the future.

**Pull request merged**